lib/query/token-visitors.js

```diff
@@ -0,0 +1,188 @@
+const { isEmpty, compact, isNil, isFunction, isString, cloneDeep, castArray } = require("lodash");
+const Tokenizer = require("tokenize-this");
+const { operators } = require("../operators");
+
+class TokenListVisitor {
+    constructor(tokens) {
+        if (isString(tokens)) tokens = this.tokenize(tokens);
+        this.tokens = tokens;
+    }
+    tokenize(rawStr) {
+        const tokens = [];
+        const tokenizer = new Tokenizer();
+        tokenizer.tokenize(rawStr, (token) => {
+            if (tokens.length > 0) {
+                for (const combined of [
+                    tokens[tokens.length - 1] + token,
+                    tokens[tokens.length - 1] + ' ' + token
+                ]) {
+                    if (operators[combined]) {
+                        tokens[tokens.length - 1] = combined;
+                        return;
+                    }
+                }
+            }
+            tokens.push(`${token}`);
+        });
+        return tokens;
+    }
+    toString() {
+        return compact(this.tokens).join(' ');
+    }
+    findToken(value, isCaseSensitive = false, startIdx = 0) {
+        const lowerValue = value.toLowerCase();
+        for (let idx = startIdx; idx < this.tokens.length; idx++) {
```

Comment:
why not juse use .find() method of JS arrays?

Reply:
Because the starting index may not be 0.